File: packages/runtime-go/core/oauth2_test.go
package core

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"sync"
	"testing"
)

// The token request is barely worth testing. These cover the three things around it that are easy to
// get wrong and expensive when wrong (SPEC.md §3.1.6): single-flight refresh, proactive expiry, and
// retrying a 401 exactly once.

func tokenServer(t *testing.T, handler func(w http.ResponseWriter, r *http.Request, call int)) (string, *int) {
	t.Helper()
	var mu sync.Mutex
	calls := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		calls++
		call := calls
		mu.Unlock()
		w.Header().Set("Content-Type", "application/json")
		handler(w, r, call)
	}))
	t.Cleanup(server.Close)
	return server.URL, &calls
}

func TestTokenRequestSendsClientCredentialsWithBasicAuth(t *testing.T) {
	var gotAuth, gotBody string
	tokenURL, _ := tokenServer(t, func(w http.ResponseWriter, r *http.Request, _ int) {
		gotAuth = r.Header.Get("Authorization")
		body := make([]byte, 512)
		n, _ := r.Body.Read(body)
		gotBody = string(body[:n])
		_, _ = w.Write([]byte(`{"access_token":"tok_1","expires_in":3600}`))
	})

	source := NewTokenSource(OAuth2Config{
		Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id", ClientSecret: "secret",
	}, nil)

	token, err := source.Token(context.Background())
	if err != nil {
		t.Fatalf("Token: %v", err)
	}
	if token != "tok_1" {
		t.Errorf("token = %q", token)
	}
	if !strings.HasPrefix(gotAuth, "Basic ") {
		t.Errorf("Authorization = %q, want the header form (RFC 6749 requires servers to accept it)", gotAuth)
	}
	if !strings.Contains(gotBody, "grant_type=client_credentials") {
		t.Errorf("body = %q", gotBody)
	}
	// The credentials must not also appear in the body.
	if strings.Contains(gotBody, "client_secret") {
		t.Error("the secret was sent in both the header and the body")
	}
}

func TestTokenRequestCanSendCredentialsInTheBody(t *testing.T) {
	var gotAuth, gotBody string
	tokenURL, _ := tokenServer(t, func(w http.ResponseWriter, r *http.Request, _ int) {
		gotAuth = r.Header.Get("Authorization")
		body := make([]byte, 512)
		n, _ := r.Body.Read(body)
		gotBody = string(body[:n])
		_, _ = w.Write([]byte(`{"access_token":"t"}`))
	})
	source := NewTokenSource(OAuth2Config{
		Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id", ClientSecret: "secret",
		ClientAuth: ClientAuthBody,
	}, nil)
	if _, err := source.Token(context.Background()); err != nil {
		t.Fatalf("Token: %v", err)
	}
	if gotAuth != "" {
		t.Errorf("Authorization = %q, want empty", gotAuth)
	}
	if !strings.Contains(gotBody, "client_id=id") {
		t.Errorf("body = %q", gotBody)
	}
}

func TestTokenRequestSendsScopesSpaceSeparated(t *testing.T) {
	var gotScope string
	tokenURL, _ := tokenServer(t, func(w http.ResponseWriter, r *http.Request, _ int) {
		body := make([]byte, 512)
		n, _ := r.Body.Read(body)
		// Parsed rather than string-matched: form encoding turns a space into `+`.
		values, _ := url.ParseQuery(string(body[:n]))
		gotScope = values.Get("scope")
		_, _ = w.Write([]byte(`{"access_token":"t"}`))
	})
	source := NewTokenSource(OAuth2Config{
		Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id",
		Scopes: []string{"read:widgets", "write:widgets"},
	}, nil)
	if _, err := source.Token(context.Background()); err != nil {
		t.Fatalf("Token: %v", err)
	}
	if gotScope != "read:widgets write:widgets" {
		t.Errorf("scope = %q", gotScope)
	}
}

func TestTokenIsReusedWhileValid(t *testing.T) {
	tokenURL, calls := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, _ int) {
		_, _ = w.Write([]byte(`{"access_token":"tok","expires_in":3600}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)
	for i := 0; i < 3; i++ {
		if _, err := source.Token(context.Background()); err != nil {
			t.Fatalf("Token: %v", err)
		}
	}
	if *calls != 1 {
		t.Errorf("made %d token requests, want 1", *calls)
	}
}

func TestTokenRefreshesBeforeItActuallyExpires(t *testing.T) {
	// Refreshing only once a token has expired guarantees at least one failed request first. A
	// lifetime shorter than the 30s skew is already expired on arrival, which is what makes this
	// observable without manipulating the clock.
	tokenURL, calls := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, call int) {
		_, _ = w.Write([]byte(`{"access_token":"tok_` + string(rune('0'+call)) + `","expires_in":10}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)
	first, err := source.Token(context.Background())
	if err != nil {
		t.Fatalf("Token: %v", err)
	}
	second, err := source.Token(context.Background())
	if err != nil {
		t.Fatalf("Token: %v", err)
	}
	if first == second || *calls != 2 {
		t.Errorf("a token inside the safety margin was reused: %q then %q after %d calls", first, second, *calls)
	}
}

func TestNoExpiresInMeansLongLived(t *testing.T) {
	tokenURL, calls := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, _ int) {
		_, _ = w.Write([]byte(`{"access_token":"tok"}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)
	for i := 0; i < 3; i++ {
		if _, err := source.Token(context.Background()); err != nil {
			t.Fatalf("Token: %v", err)
		}
	}
	// Nothing to expire against, so nothing to refresh.
	if *calls != 1 {
		t.Errorf("made %d token requests, want 1", *calls)
	}
}

func TestSingleFlightMakesOneRequestForManyGoroutines(t *testing.T) {
	// Without this, the first thing a new SDK does under load is hammer the authorization server, and
	// the symptom is unexplained 429s from a host the caller never configured.
	release := make(chan struct{})
	tokenURL, calls := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, _ int) {
		<-release
		_, _ = w.Write([]byte(`{"access_token":"tok","expires_in":3600}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)

	var wg sync.WaitGroup
	tokens := make([]string, 10)
	errs := make([]error, 10)
	for i := 0; i < 10; i++ {
		wg.Add(1)
		go func(index int) {
			defer wg.Done()
			tokens[index], errs[index] = source.Token(context.Background())
		}(i)
	}
	close(release)
	wg.Wait()

	if *calls != 1 {
		t.Errorf("made %d token requests, want 1", *calls)
	}
	for i, err := range errs {
		if err != nil {
			t.Fatalf("goroutine %d: %v", i, err)
		}
		if tokens[i] != "tok" {
			t.Errorf("goroutine %d got %q", i, tokens[i])
		}
	}
}

func TestAFailedFetchDoesNotPoisonLaterCalls(t *testing.T) {
	// A permanently cached failure — which sync.Once would give — would make every later call fail
	// with the same stale error long after the cause was fixed.
	var mu sync.Mutex
	failing := true
	tokenURL, _ := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, _ int) {
		mu.Lock()
		shouldFail := failing
		mu.Unlock()
		if shouldFail {
			w.WriteHeader(http.StatusServiceUnavailable)
			_, _ = w.Write([]byte(`{"error":"temporarily_unavailable"}`))
			return
		}
		_, _ = w.Write([]byte(`{"access_token":"tok"}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)

	if _, err := source.Token(context.Background()); err == nil {
		t.Fatal("expected the first fetch to fail")
	}
	mu.Lock()
	failing = false
	mu.Unlock()
	token, err := source.Token(context.Background())
	if err != nil {
		t.Fatalf("the second fetch inherited the first failure: %v", err)
	}
	if token != "tok" {
		t.Errorf("token = %q", token)
	}
}

func TestRefreshTokenFlowAdoptsARotatedToken(t *testing.T) {
	// A server that rotates refresh tokens invalidates the old one, so reusing it would fail on the
	// *second* refresh — a bug that only appears after a token lifetime has elapsed.
	var sent []string
	tokenURL, _ := tokenServer(t, func(w http.ResponseWriter, r *http.Request, call int) {
		body := make([]byte, 512)
		n, _ := r.Body.Read(body)
		values, _ := url.ParseQuery(string(body[:n]))
		sent = append(sent, values.Get("refresh_token"))
		n2 := string(rune('0' + call))
		_, _ = w.Write([]byte(`{"access_token":"a_` + n2 + `","refresh_token":"r_` + n2 + `","expires_in":10}`))
	})
	source := NewTokenSource(OAuth2Config{
		Flow: FlowRefreshToken, TokenURL: tokenURL, RefreshToken: "r_0",
	}, nil)

	if _, err := source.Token(context.Background()); err != nil {
		t.Fatalf("Token: %v", err)
	}
	if _, err := source.Token(context.Background()); err != nil {
		t.Fatalf("Token: %v", err)
	}
	if len(sent) != 2 || sent[0] != "r_0" || sent[1] != "r_1" {
		t.Errorf("refresh tokens sent = %v, want [r_0 r_1]", sent)
	}
}

func TestTokenFailureSurfacesTheServerCode(t *testing.T) {
	tokenURL, calls := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, _ int) {
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`{"error":"invalid_client","error_description":"bad secret"}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)
	_, err := source.Token(context.Background())

	var oauthErr *OAuth2Error
	if !errors.As(err, &oauthErr) {
		t.Fatalf("expected *OAuth2Error, got %T", err)
	}
	if oauthErr.Code != "invalid_client" || oauthErr.Message != "bad secret" || oauthErr.Status != 401 {
		t.Errorf("error = %+v", oauthErr)
	}
	// Never retried: a 400-class response means the credentials are wrong.
	if *calls != 1 {
		t.Errorf("made %d token requests, want 1", *calls)
	}
	// Not an APIError, so a caller handling a 4xx from the API does not catch it by accident.
	if _, ok := AsAPIError(err); ok {
		t.Error("a token failure presented as an APIError")
	}
	// Still an SDKError, so `errors.As(err, &sdkErr)` catches everything the SDK returns.
	var sdkErr SDKError
	if !errors.As(err, &sdkErr) {
		t.Error("a token failure is not an SDKError")
	}
}

func TestTokenFailureOnMissingAccessToken(t *testing.T) {
	tokenURL, _ := tokenServer(t, func(w http.ResponseWriter, _ *http.Request, _ int) {
		_, _ = w.Write([]byte(`{"token_type":"Bearer"}`))
	})
	source := NewTokenSource(OAuth2Config{Flow: FlowClientCredentials, TokenURL: tokenURL, ClientID: "id"}, nil)
	_, err := source.Token(context.Background())
	if err == nil || !strings.Contains(err.Error(), "no access_token") {
		t.Errorf("err = %v", err)
	}
}

// --- integration with the client's retry loop ---

func oauthClient(t *testing.T, apiStatuses []int) (*Client, *[]string) {
	t.Helper()
	var mu sync.Mutex
	var log []string
	tokens, apis := 0, 0

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		defer mu.Unlock()
		w.Header().Set("Content-Type", "application/json")
		if strings.Contains(r.URL.Path, "/token") {
			tokens++
			log = append(log, "token")
			_, _ = w.Write([]byte(`{"access_token":"tok"}`))
			return
		}
		status := http.StatusOK
		if apis < len(apiStatuses) {
			status = apiStatuses[apis]
		}
		apis++
		log = append(log, "api:"+http.StatusText(status))
		w.WriteHeader(status)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	t.Cleanup(server.Close)

	source := NewTokenSource(OAuth2Config{
		Flow: FlowClientCredentials, TokenURL: server.URL + "/token", ClientID: "id",
	}, server.Client())
	client := NewClient(ClientOptions{
		BaseURL:    server.URL,
		Auth:       Auth{Kind: AuthOAuth2, Source: source},
		MaxRetries: -1,
		HTTPClient: server.Client(),
	})
	return client, &log
}

func TestA401ForcesOneRefreshAndRetriesOnce(t *testing.T) {
	// Clocks disagree and servers revoke tokens early, so expiry arithmetic is necessary but not
	// sufficient.
	client, log := oauthClient(t, []int{http.StatusUnauthorized, http.StatusOK})
	var out map[string]any
	if err := client.DoJSON(context.Background(), &Request{Method: "get", Path: "/widgets"}, &out); err != nil {
		t.Fatalf("DoJSON: %v", err)
	}
	want := []string{"token", "api:Unauthorized", "token", "api:OK"}
	if strings.Join(*log, ",") != strings.Join(want, ",") {
		t.Errorf("log = %v, want %v", *log, want)
	}
}

func TestA401GivesUpAfterOneRefresh(t *testing.T) {
	// A genuinely revoked credential must not become a loop against the authorization server.
	client, log := oauthClient(t, []int{http.StatusUnauthorized, http.StatusUnauthorized, http.StatusUnauthorized})
	var out map[string]any
	if err := client.DoJSON(context.Background(), &Request{Method: "get", Path: "/widgets"}, &out); err == nil {
		t.Fatal("expected an error")
	}
	tokenRequests := 0
	for _, entry := range *log {
		if entry == "token" {
			tokenRequests++
		}
	}
	if tokenRequests != 2 {
		t.Errorf("made %d token requests, want 2 (one initial, one forced)", tokenRequests)
	}
}

func TestANon401DoesNotForceARefresh(t *testing.T) {
	client, log := oauthClient(t, []int{http.StatusForbidden})
	var out map[string]any
	if err := client.DoJSON(context.Background(), &Request{Method: "get", Path: "/widgets"}, &out); err == nil {
		t.Fatal("expected an error")
	}
	tokenRequests := 0
	for _, entry := range *log {
		if entry == "token" {
			tokenRequests++
		}
	}
	if tokenRequests != 1 {
		t.Errorf("made %d token requests, want 1", tokenRequests)
	}
}

File: packages/runtime-typescript/src/oauth2.test.ts
/**
 * OAuth2 tests.
 *
 * The token request itself is barely worth testing. What these cover is the three things around it
 * that are easy to get wrong and expensive when wrong (SPEC.md §3.1.6): single-flight refresh,
 * proactive expiry, and retrying a 401 exactly once.
 */

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { BaseClient } from './client.js';
import { APIError } from './errors.js';
import { OAuth2Error, TokenSource } from './oauth2.js';

interface Recorded {
  readonly url: string;
  readonly headers: Record<string, string>;
  readonly body: string;
}

function tokenServer(
  respond: (call: number) => { status?: number; body: unknown },
): { fetchImpl: typeof globalThis.fetch; calls: Recorded[] } {
  const calls: Recorded[] = [];
  const impl = vi.fn(async (input: string | URL | Request, init?: RequestInit) => {
    calls.push({
      url: String(input),
      headers: Object.fromEntries(
        Object.entries((init?.headers ?? {}) as Record<string, string>).map(([k, v]) => [
          k.toLowerCase(),
          v,
        ]),
      ),
      body: String(init?.body ?? ''),
    });
    const { status = 200, body } = respond(calls.length);
    return new Response(JSON.stringify(body), {
      status,
      headers: { 'content-type': 'application/json' },
    });
  });
  return { fetchImpl: impl as unknown as typeof globalThis.fetch, calls };
}

const clientCredentials = {
  flow: 'clientCredentials',
  tokenUrl: 'https://auth.test/token',
  clientId: 'id',
  clientSecret: 'secret',
} as const;

beforeEach(() => vi.useFakeTimers());
afterEach(() => vi.useRealTimers());

describe('the token request', () => {
  it('sends client_credentials with basic client authentication', async () => {
    const { fetchImpl, calls } = tokenServer(() => ({
      body: { access_token: 'tok_1', expires_in: 3600 },
    }));
    const source = new TokenSource(clientCredentials, fetchImpl);

    expect(await source.token()).toBe('tok_1');
    expect(calls[0]!.url).toBe('https://auth.test/token');
    expect(calls[0]!.body).toContain('grant_type=client_credentials');
    // RFC 6749 requires servers to accept the header form, so it is the default.
    expect(calls[0]!.headers.authorization).toBe(`Basic ${btoa('id:secret')}`);
    // The credentials must not also appear in the body.
    expect(calls[0]!.body).not.toContain('client_secret');
  });

  it('can send credentials in the body for servers that require it', async () => {
    const { fetchImpl, calls } = tokenServer(() => ({ body: { access_token: 't' } }));
    const source = new TokenSource({ ...clientCredentials, clientAuth: 'body' }, fetchImpl);
    await source.token();
    expect(calls[0]!.headers.authorization).toBeUndefined();
    expect(calls[0]!.body).toContain('client_id=id');
  });

  it('sends requested scopes space-separated', async () => {
    const { fetchImpl, calls } = tokenServer(() => ({ body: { access_token: 't' } }));
    await new TokenSource(
      { ...clientCredentials, scopes: ['read:widgets', 'write:widgets'] },
      fetchImpl,
    ).token();
    // Parsed rather than string-matched: URLSearchParams encodes a space as `+`, which
    // decodeURIComponent does not undo — so a naive assertion tests the encoding, not the value.
    expect(new URLSearchParams(calls[0]!.body).get('scope')).toBe('read:widgets write:widgets');
  });
});

describe('caching and proactive expiry', () => {
  it('reuses a token that is still valid', async () => {
    const { fetchImpl, calls } = tokenServer(() => ({
      body: { access_token: 'tok', expires_in: 3600 },
    }));
    const source = new TokenSource(clientCredentials, fetchImpl);
    await source.token();
    await source.token();
    await source.token();
    expect(calls).toHaveLength(1);
  });

  it('refreshes before the token actually expires', async () => {
    // Refreshing only once a token has expired guarantees at least one failed request first.
    const { fetchImpl, calls } = tokenServer((n) => ({
      body: { access_token: `tok_${n}`, expires_in: 60 },
    }));
    const source = new TokenSource(clientCredentials, fetchImpl);
    expect(await source.token()).toBe('tok_1');

    // 35s in: still inside the 60s lifetime, but past the 30s safety margin.
    vi.setSystemTime(Date.now() + 35_000);
    expect(await source.token()).toBe('tok_2');
    expect(calls).toHaveLength(2);
  });

  it('treats a token with no expires_in as long-lived', async () => {
    const { fetchImpl, calls } = tokenServer(() => ({ body: { access_token: 'tok' } }));
    const source = new TokenSource(clientCredentials, fetchImpl);
    await source.token();
    vi.setSystemTime(Date.now() + 86_400_000);
    await source.token();
    // Nothing to expire against, so nothing to refresh. A server that means otherwise must say so.
    expect(calls).toHaveLength(1);
  });
});

describe('single-flight refresh', () => {
  it('makes one token request for many concurrent callers', async () => {
    // Without this, the first thing a new SDK does under load is hammer the authorization server, and
    // the symptom is unexplained 429s from a host the caller never configured.
    let release: (() => void) | undefined;
    const gate = new Promise<void>((resolve) => {
      release = resolve;
    });
    const calls: string[] = [];
    const fetchImpl = (async () => {
      calls.push('token');
      await gate;
      return new Response(JSON.stringify({ access_token: 'tok', expires_in: 3600 }), {
        headers: { 'content-type': 'application/json' },
      });
    }) as unknown as typeof globalThis.fetch;

    const source = new TokenSource(clientCredentials, fetchImpl);
    const pending = Promise.all(Array.from({ length: 10 }, () => source.token()));
    release!();
    const tokens = await pending;

    expect(calls).toHaveLength(1);
    expect(new Set(tokens)).toEqual(new Set(['tok']));
  });

  it('does not poison later calls when a refresh fails', async () => {
    // A rejected in-flight promise left in place would make every subsequent call fail with the same
    // stale rejection, long after the cause was fixed.
    let failing = true;
    const fetchImpl = (async () => {
      if (failing) return new Response('{"error":"temporarily_unavailable"}', { status: 503 });
      return new Response(JSON.stringify({ access_token: 'tok' }), {
        headers: { 'content-type': 'application/json' },
      });
    }) as unknown as typeof globalThis.fetch;

    const source = new TokenSource(clientCredentials, fetchImpl);
    await expect(source.token()).rejects.toBeInstanceOf(OAuth2Error);
    failing = false;
    expect(await source.token()).toBe('tok');
  });
});

describe('refresh-token flow', () => {
  it('sends the refresh grant and adopts a rotated token', async () => {
    // A server that rotates refresh tokens invalidates the old one, so reusing it would fail on the
    // second refresh.
    const { fetchImpl, calls } = tokenServer((n) => ({
      body: { access_token: `a_${n}`, refresh_token: `r_${n}`, expires_in: 60 },
    }));
    const source = new TokenSource(
      { flow: 'refreshToken', tokenUrl: 'https://auth.test/token', refreshToken: 'r_0' },
      fetchImpl,
    );

    expect(await source.token()).toBe('a_1');
    expect(calls[0]!.body).toContain('refresh_token=r_0');

    vi.setSystemTime(Date.now() + 35_000);
    expect(await source.token()).toBe('a_2');
    expect(calls[1]!.body).toContain('refresh_token=r_1');
  });
});

describe('failures', () => {
  it('surfaces the server error code and description', async () => {
    const { fetchImpl } = tokenServer(() => ({
      status: 401,
      body: { error: 'invalid_client', error_description: 'Client authentication failed' },
    }));
    const source = new TokenSource(clientCredentials, fetchImpl);
    await expect(source.token()).rejects.toMatchObject({
      name: 'OAuth2Error',
      status: 401,
      code: 'invalid_client',
      message: 'Client authentication failed',
    });
  });

  it('never retries a token request', async () => {
    // A 400 from a token endpoint means the credentials are wrong. Retrying is pointless and
    // indistinguishable from a brute-force attempt.
    const { fetchImpl, calls } = tokenServer(() => ({ status: 400, body: { error: 'invalid_grant' } }));
    const source = new TokenSource(clientCredentials, fetchImpl);
    await expect(source.token()).rejects.toBeInstanceOf(OAuth2Error);
    expect(calls).toHaveLength(1);
  });

  it('reports a response with no access_token', async () => {
    const { fetchImpl } = tokenServer(() => ({ body: { token_type: 'Bearer' } }));
    await expect(new TokenSource(clientCredentials, fetchImpl).token()).rejects.toThrowError(
      /no access_token/,
    );
  });

  it('reports an unreachable token endpoint', async () => {
    const fetchImpl = (async () => {
      throw new TypeError('network down');
    }) as unknown as typeof globalThis.fetch;
    await expect(new TokenSource(clientCredentials, fetchImpl).token()).rejects.toMatchObject({
      name: 'OAuth2Error',
      status: 0,
    });
  });
});

describe('a 401 from the API', () => {
  /** A fetch that answers the token endpoint and the API from one function. */
  function combined(apiStatuses: number[]): { fetchImpl: typeof globalThis.fetch; log: string[] } {
    const log: string[] = [];
    let tokenCount = 0;
    let apiCount = 0;
    const impl = (async (input: string | URL | Request) => {
      const url = String(input);
      if (url.includes('/token')) {
        tokenCount += 1;
        log.push(`token:${tokenCount}`);
        return new Response(JSON.stringify({ access_token: `tok_${tokenCount}` }), {
          headers: { 'content-type': 'application/json' },
        });
      }
      const status = apiStatuses[apiCount] ?? 200;
      apiCount += 1;
      log.push(`api:${status}`);
      return new Response(JSON.stringify({ ok: status === 200 }), {
        status,
        headers: { 'content-type': 'application/json' },
      });
    }) as unknown as typeof globalThis.fetch;
    return { fetchImpl: impl, log };
  }

  function client(fetchImpl: typeof globalThis.fetch): BaseClient {
    const source = new TokenSource(clientCredentials, fetchImpl);
    return new BaseClient({
      baseURL: 'https://api.test',
      auth: { type: 'oauth2', source },
      maxRetries: 0,
      fetch: fetchImpl,
    });
  }

  it('forces one refresh and retries once', async () => {
    // Clocks disagree and servers revoke tokens early, so expiry arithmetic is necessary but not
    // sufficient.
    const { fetchImpl, log } = combined([401, 200]);
    const result = await client(fetchImpl).request<{ ok: boolean }>({
      method: 'get',
      path: '/widgets',
      responseType: 'json',
    });
    expect(result).toEqual({ ok: true });
    expect(log).toEqual(['token:1', 'api:401', 'token:2', 'api:200']);
  });

  it('gives up after one refresh rather than looping', async () => {
    // A genuinely revoked credential must not become a loop against the authorization server.
    const { fetchImpl, log } = combined([401, 401, 401]);
    await expect(
      client(fetchImpl).request({ method: 'get', path: '/widgets', responseType: 'json' }),
    ).rejects.toBeInstanceOf(APIError);
    expect(log.filter((entry) => entry.startsWith('token:'))).toHaveLength(2);
    expect(log.filter((entry) => entry === 'api:401')).toHaveLength(2);
  });

  it('does not refresh for a non-401 failure', async () => {
    const { fetchImpl, log } = combined([403]);
    await expect(
      client(fetchImpl).request({ method: 'get', path: '/widgets', responseType: 'json' }),
    ).rejects.toBeInstanceOf(APIError);
    expect(log.filter((entry) => entry.startsWith('token:'))).toHaveLength(1);
  });
});
